apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/TransactionPool.cs

```diff
@@ -0,0 +1,247 @@
+{
+    /// <summary>
+    /// The TransactionPool works in conjuntion with the Session pool to warm pooled Sessions
+    /// with the transaction semantics it last used (if appropriate).
+    /// When requesting a pooled session, the request can include transactionoptions which
+    /// is then used as a key in the sessionpool.
+    /// </summary>
+    public static class TransactionPool
+    {
+        // Holds transaction state on sessions, including the last created transaction and options used.
+        static readonly ConcurrentDictionary<Session, SessionInfo> s_sessionInfoTable = new ConcurrentDictionary<Session, SessionInfo>();
+
+        // Holds a list of all active transactions with back pointers to the owning session.
+        static readonly ConcurrentDictionary<ByteString, Session> s_activeTransactionTable = new ConcurrentDictionary<ByteString, Session>();
+
+        /// <summary>
+        /// Creates a transaction with the given transactoin options.
+        /// Depending on the state of the session, this may or may not result in a request
+        /// to Spanner over the network.
+        /// </summary>
+        /// <param name="session"></param>
+        /// <param name="client"></param>
+        /// <param name="options"></param>
+        /// <returns></returns>
+        public static async Task<Transaction> BeginPooledTransactionAsync(this SpannerClient client, Session session,
+            TransactionOptions options)
+        {
+            var info = s_sessionInfoTable.GetOrAdd(session, s => new SessionInfo {SpannerClient = client});
+        
+            //we need to await for previous task completion anyway -- otherwise there is a bad race condition.
+            await info.WaitForPreWarmAsync().ConfigureAwait(false);
+            
+            //now we see if we can return the prewarmed transaction.
+            // a lot has to be correct for us to return the precreated transaction.
+            // we have to have a valid transaction object.
+            // it needs to exist in the activetransactiontable to indicate that it hasn't been closed
+            // the options on it must equal the passed in options.
+            if (info.ActiveTransaction != null
+                && !info.ActiveTransaction.Id.IsEmpty
+                && s_activeTransactionTable.ContainsKey(info.ActiveTransaction.Id))
+            {
+                if (info.ActiveTransactionOptions != null
+                    && info.ActiveTransactionOptions.Equals(options))
+                {
+                    return info.ActiveTransaction;
+                }
+                //cache hit, but no match on options
+                Session ignored;
+                s_activeTransactionTable.TryRemove(info.ActiveTransaction.Id, out ignored);
+            }
+
+            //ok, our cache hit didnt work for whatever reason.  Let's create a transaction with the given options and return it.
+            await info.CreateTransactionAsync(session, options);
```

Comment:
agree we'll miss some.  Absolutely mandatory testing will be to ensure every async call can be called on the main UI thread without deadlocking.  This will at least capture the most important ones.
You can argue that the inner awaits aren't as important as long as they are already running on a threadpool thread (ie configureawait has already been previously called), right?